File: templates/problems/11_graphs/10_graph_valid_tree/typescript/solution.ts
class UnionFind {
    parent: number[];
    rank: number[];

    constructor(n: number) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = new Array(n).fill(0);
    }

    find(x: number): number {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x: number, y: number): boolean {
        let px = this.find(x);
        let py = this.find(y);
        if (px === py) return false;
        if (this.rank[px] < this.rank[py]) {
            [px, py] = [py, px];
        }
        this.parent[py] = px;
        if (this.rank[px] === this.rank[py]) this.rank[px]++;
        return true;
    }
}

/**
 * Check if the graph is a valid tree using Union-Find.
 * Time: O(n * alpha(n)) ~= O(n), Space: O(n).
 */
export function solveValidTree(n: number, edges: number[][]): boolean {
    if (edges.length !== n - 1) return false;
    const uf = new UnionFind(n);
    for (const e of edges) {
        if (!uf.union(e[0], e[1])) return false;
    }
    return true;
}

File: templates/problems/11_graphs/10_graph_valid_tree/python/solution.py
class UnionFind:
    """Union-Find (Disjoint Set) with path compression and union by rank."""

    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        px, py = self.find(x), self.find(y)
        if px == py:
            return False
        if self.rank[px] < self.rank[py]:
            px, py = py, px
        self.parent[py] = px
        if self.rank[px] == self.rank[py]:
            self.rank[px] += 1
        return True


def solve_valid_tree(n: int, edges: list[list[int]]) -> bool:
    """Check if the graph is a valid tree using Union-Find.

    Time: O(n * alpha(n)) ~= O(n), Space: O(n).
    """
    if len(edges) != n - 1:
        return False
    uf = UnionFind(n)
    for e in edges:
        if not uf.union(e[0], e[1]):
            return False
    return True
